scipy/_lib/_array_api.py

```diff
@@ -260,6 +261,17 @@ def default_xp(xp: ModuleType) -> Generator[None, None, None]:
         _default_xp_ctxvar.reset(token)
 
 
+def eager_warns(x, warning_type, match=None):
+    """pytest.warns context manager, but only if x is not a lazy array."""
+    import pytest
+    # This attribute is interpreted by pytest-run-parallel, ensuring that tests that use
+    # `eager_warns` aren't run in parallel (since pytest.warns isn't thread-safe).
+    __thread_safe__ = False  # noqa: F841
+    if is_lazy_array(x):
+        return contextlib.nullcontext()
```

Comment:
@rgommers So if we were to change the signature of `eager_warns` to accept `xp` instead of an array, we'd either change this to:

```suggestion:
    if is_lazy_array(xp.asarray(0)):
        return contextlib.nullcontext()
```

or we'd want an `is_lazy_namespace` function that would accept `xp`.

Would this function go in `array-api-compat`, or would it be private in SciPy? I thought that maybe `array-api-compat` would refuse to define an `is_lazy_namespace` function in case arrays from a namespace could be either lazy or eager. But according to the [documentation](https://data-apis.org/array-api-compat/helper-functions.html#array_api_compat.is_lazy_array):

> This function errs on the side of caution for array types that may or may not be lazy, e.g. JAX arrays, by always returning True for them.

This suggests that there might as well be an `is_lazy_namespace` function that also "errs on the side of caution". If using on an unrecognized namespace, it could either raise, or it could call `is_lazy_array(xp.asarray(0))` or similar.

Reply:
Ah, great question. That does seem a little problematic for libraries that have both lazy and eager flavors. Looking at the current implementation, so far we have had no need to deal with `torch.compile` or other PyTorch backends as lazy, but that could easily change - there are fully lazy backends there after all.

I think the best choice is to leave `eager_warns` as is, and not change it to using `xp=xp`. The only upside of the change is consistency in signature with other functions, but this question above shows that the design is better off accepting an array rather than a namespace.

@crusaderky may have things to add here, since he originally wrote this function - but I suspect he considered both options.

Reply:
At the moment, torch is always eager according to `is_lazy_array`, but that may change pretty soon in the future.
JAX is always treated as lazy by design (eager JAX is considered a development and debug tool of jitted JAX).

Another backend which can be eager or lazy is blosc2. Under the hood, it lazifies chained operations with `numexpr`. It's been off our radar so far due to trivial issues (`blosc2.asarray` doesn't accept an ArrayLike) but I expect it to join the fray pretty soon.

```python
>>>x = blosc2.asarray(np.asarray([1,2,3]))
>>> x
<blosc2.ndarray.NDArray at 0x7b3d2f461900>
>>> x + 1
<blosc2.lazyexpr.LazyExpr at 0x7b3e5bfbb6a0>
```

Reply:
If `is_lazy_array` will return different results for a given backend in the near future, it sounds fine to leave the `array-api-compat` function alone.

As for the tests, I would prefer the more familiar interface if we can almost always make the distinction based on `xp`. I'd even volunteer to add a third possibility for the context that *ignores* the specified warning (rather than failing if it is emitted or failing if it is not emitted) if we're comfortable with the slight loss in test coverage. Given that the functions being tested will be using the arrays to make the distinction with `is_lazy_array`, it seems unlikely that the warning would not be emitted correctly based on the laziness of the array - and unimportant, since warning behavior for such backends will probably seem inconsistent to users anyway.

Reply:
Cool. @rgommers does that work for you - assert that the warning is emitted when backends are inherently eager, and ignore the warnings otherwise? If that's OK, I can modify `eager_warns` to accept `xp` by keyword instead of an array as the first argument without changing `is_lazy_array` - it would just manually specify the context for each backend we test.

Reply:
> ignoring the warning instead of asserting that there is none sounds good to me.

This SGTM too.

> (Do we need to test that warnings are not emited for lazy backends, or would there be another failure if the function attempted to warn for a lazy backend?)

None of that I think, that'll just give us a bunch of extra fragility without any upside I'd expect.

I'd be fine with testing eager-only warnings and leaving out the rest completely, or even only testing numpy-only warnings.